src/commands/classified/external.rs

```diff
@@ -50,158 +49,243 @@ impl Decoder for LinesCodec {
+            || (arg_chars.len() > 1
+                && arg_chars[0] == '\''
+                && arg_chars[arg_chars.len() - 1] == '\'')
+        {
+            // quoted string
+            let new_arg: String = arg_chars[1..arg_chars.len() - 1].iter().collect();
+            process = process.arg(new_arg);
+        } else {
+            process = process.arg(arg.as_ref());
         }
-    };
+    }
 
-    trace!(target: "nu::run::external", "set up stdout pipe");
+    process = process.cwd(context.shell_manager.path());
+    trace!(target: "nu::run::external", "cwd = {:?}", context.shell_manager.path());
 
-    if let Some(stdin) = stdin {
-        process = process.stdin(stdin);
+    if !is_last {
+        process = process.stdout(subprocess::Redirection::Pipe);
+        trace!(target: "nu::run::external", "set up stdout pipe");
     }
 
+    // NOTE can't use a subprocess::make_pipe here. There seems to be no way to close stdin
+    //      after doing so, which makes certain commands block forever (e.g., cat).
+    process = process.stdin(subprocess::Redirection::Pipe);
     trace!(target: "nu::run::external", "set up stdin pipe");
+
     trace!(target: "nu::run::external", "built process {:?}", process);
 
     let popen = process.popen();
+    if let Ok(mut popen) = popen {
+        let _ = popen.detach();
+        let mut stdin_write = popen.stdin.take().unwrap();
 
-    trace!(target: "nu::run::external", "next = {:?}", stream_next);
+        // Instead of writing everything into the pipe, which may(?) overflow, we write small
+        // amounts at a time.
+        let name_tag2 = name_tag.clone();
+        let write_to_stdin_stream = async_stream! {
+            while let Some(item) = input.next().await {
+                match item.value {
+                    UntaggedValue::Primitive(Primitive::String(s)) => {
+                        stdin_write.write(s.as_bytes());
+                        stdin_write.write(b"\n");
+                        yield Ok(());
+                    },
+                    UntaggedValue::Primitive(Primitive::Line(line)) => {
+                        stdin_write.write(line.as_bytes());
+                        yield Ok(());
+                    },
 
-    let name_tag = command.name_tag.clone();
-    if let Ok(mut popen) = popen {
-        match stream_next {
-            StreamNext::Last => {
-                popen.detach();
+                    // TODO perhaps serialize some other primitives?
+                    _ => {
+                        yield Err(ShellError::labeled_error(
+                            "Unexpected stdin type found in input stream",
+                            "unexpected type",
+                            name_tag2.clone()
+                        ));
+                        break;
+                    }
+                }
+            }
+        };
+
+        let stream = async_stream! {
+            pin_mut!(write_to_stdin_stream);
```

Comment:
Is it easier to make this two async_stream steps rather than one async_stream that does the job of pulling on input and outputting to stdin?  I can believe that makes it too complicated, just curious

Reply:
My thought was that coordinating would make it less likely that we fill up a pipe. That being said, I guess once it's detached and processing, we can probably write into it as quickly as possible 🤔 

Yeah, I think that makes sense, and will likely be better performance, since we just write as quickly as possible 👍